torch/cuda/__init__.py

```diff
@@ -181,16 +195,19 @@ def _lazy_init():
     _cudart.cudaGetErrorName.restype = ctypes.c_char_p
     _cudart.cudaGetErrorString.restype = ctypes.c_char_p
     _original_pid = os.getpid()
-    _initialized = True
-    # Important to do this after _initialized, since some queued calls
-    # may themselves call _lazy_init()
+    # Some of the queued calls may reentrantly call _lazy_init();
+    # we need to just return without initializing in that case.
+    # However, we must not let any *other* threads in!
+    _tls.is_initializing = True
     for queued_call, orig_traceback in _queued_calls:
         try:
             queued_call()
         except Exception as e:
             msg = ("CUDA call failed lazily at initialization with error: {}\n\n"
                    "CUDA call was originally invoked at:\n\n{}").format(str(e), orig_traceback)
             raise_from(DeferredCudaCallError(msg), e)
+    _initialized = True
+    _initialization_lock.release()
```

Comment:
if dequeing the `_queued_calls` raises an exception for whatever reason, you never get to the `.release()` line, does this have bad implications?

Reply:
All other threads / cuda calls will deadlock on `_lazy_init`. I was shocked that threading.Lock didn't have a context manager version...